### PR TITLE
Fix the Android build

### PR DESCRIPTION
**Pull request by jakepetroules**

### CODE REVIEW

File: Sources/TSCBasic/misc.swift [hunk 1]
```diff
@@ -328,6 +328,18 @@ extension SystemError: CustomStringConvertible {
             while cap <= 16 * 1024 {
                 var buf = [Int8](repeating: 0, count: cap)
                 let err = TSCLibc.strerror_r(errno, &buf, buf.count)
+                #if os(Android)
```
Comment (jakepetroules) [inline Sources/TSCBasic/misc.swift:331]: @finagolfin Does this look reasonable?

**Comment by jakepetroules** [comment]
@swift-ci please smoke test

**Comment by finagolfin** [comment]
What is this about, are you getting compilation issues all of a sudden? Because [my Android CI just cross-compiled the latest Nov. 14 trunk snapshot of SwiftPM and TSC yesterday without a problem](https://github.com/finagolfin/swift-android-sdk/actions/runs/11868882848/job/33078637854#step:9:14121), after applying [these basic patches that were mostly just upstreamed](https://github.com/finagolfin/swift-android-sdk/blob/cb93bf31adeefba44ce8041a90f6c88f7241ae2d/swift-android-ci.patch#L89) with your approval.

I didn't know what to make of that linked Android overlay patch for TSC, as I didn't need it when building natively on Android last month or when cross-compiling the latest Termux 6.1 toolchain.

File: Sources/TSCBasic/misc.swift [hunk 2]
```diff
@@ -328,6 +328,18 @@ extension SystemError: CustomStringConvertible {
...
+                if String(cString: err) == strerror(EINVAL) {
+                    return "Unknown error \(errno)"
```
Comment (compnerd) [inline Sources/TSCBasic/misc.swift:333]: Is this an unknown error or is it as the name suggests, an invalid parameter?

File: Sources/TSCBasic/misc.swift [hunk 3]
```diff
@@ -328,6 +328,18 @@ extension SystemError: CustomStringConvertible {
...
+                }
+                if String(cString: err) == strerror(ERANGE) {
+                    cap *= 2
+                    continue
+                }
+                if err != nil {
```
Comment (compnerd) [inline Sources/TSCBasic/misc.swift:339]: Does `if let err` seem better?

**Comment by jakepetroules** [comment]
> What is this about, are you getting compilation issues all of a sudden? Because [my Android CI just cross-compiled the latest Nov. 14 trunk snapshot of SwiftPM and TSC yesterday without a problem](https://github.com/finagolfin/swift-android-sdk/actions/runs/11868882848/job/33078637854#step:9:14121), after applying [these basic patches that were mostly just upstreamed](https://github.com/finagolfin/swift-android-sdk/blob/cb93bf31adeefba44ce8041a90f6c88f7241ae2d/swift-android-ci.patch#L89) with your approval.
> 
> I didn't know what to make of that linked Android overlay patch for TSC, as I didn't need it when building natively on Android last month or when cross-compiling the latest Termux 6.1 toolchain.

Yeah, Android builds are using the GNU-specific definition of [strerror_r](https://linux.die.net/man/3/strerror_r) so I got a compilation failure here.

This was with the Swift 6.0.2 release and your Android SDK from https://github.com/finagolfin/swift-android-sdk/releases/tag/6.0.2, with top-of-tree main branch of SwiftTSC.

Using:

```swift
/Library/Developer/Toolchains/swift-6.0.2-RELEASE.xctoolchain/usr/bin/swift build --swift-sdk aarch64-unknown-linux-android24
```

**Comment by finagolfin** [comment]
> Android builds are using the GNU-specific definition of [strerror_r](https://linux.die.net/man/3/strerror_r) so I got a compilation failure here.

OK, this is a well-known issue with the Android build of this repo, which is remedied by [passing in `-Xswiftc -Xcc -Xswiftc -U_GNU_SOURCE` to SwiftPM](https://github.com/swiftlang/swift-package-manager/blob/cb829d83d016beb6607fde42136a079c37297d90/Utilities/bootstrap#L820), as you can see in [the CMake config here too](https://github.com/swiftlang/swift-tools-support-core/blob/91d1a0b70e36d2557229af2655d5a041a02324a8/Sources/TSCBasic/CMakeLists.txt#L56).

Unfortunately, while [this SwiftPM package manifest can add such defines easily](https://github.com/swiftlang/swift-tools-support-core/blob/5b130e04cc939373c4713b91704b0c47ceb36170/Package.swift#L65), it could not _undefine_ them without resorting to unsafe flags: maybe that should be fixed in SwiftPM.

After adding those flags and my linked Android overlay patch for this repo, I was able to cross-compile this repo for Android, then adding `--build-tests -Xlinker -landroid-spawn` allowed me to cross-compile its tests too (Android API 24 did not support the `posix_spawn` APIs yet, so I ship a compatibility library in my SDK bundle for that).

If you have any ideas on how to better handle these matters, I'm all ears.

**Comment by jakepetroules** [comment]
Do you know where that `_GNU_SOURCE` definition is originating from?

**Comment by jakepetroules** [comment]
I reverted the misc.swift changes for now, but I think we do still need the added import in Tracing.swift as I get a failure to find `getpid` otherwise.

**Comment by jakepetroules** [comment]
@swift-ci please smoke test

**Comment by finagolfin** [comment]
> Do you know where that _GNU_SOURCE definition is originating from?

The Swift compiler started [defining it for Android for some extra math function declarations six years ago](https://github.com/swiftlang/swift/pull/19862/files#diff-0062ad3fe872d09412bd3e7c72610b5ad0bd39667ebc00695351ef93c7c6815fR538). We briefly discussed removing it years ago, but figured there were likely already Swift codebases relying on it, so left it in.

**Review by finagolfin — APPROVED** [review]
Thanks, I confirmed that this fixed cross-compiling for Android.

**Comment by jakepetroules** [comment]
@swift-ci please smoke test

**Comment by owenv** [comment]
Looks reasonable, especially now that it's just adding the missing import

**Review by owenv — APPROVED** [review]

**Comment by jakepetroules** [comment]
@swift-ci please test Linux

**Comment by jakepetroules** [comment]
Here's a better attempt at the strerror thing which allows the package build to work regardless of whether _GNU_SOURCE is given on the command line: #497

**Pull request merged**